File: python/src/arker/migrate.py
"""Command migration: move a running command (by PID) into an Arker VM.

This is the config-driven engine — every command-specific detail lives in
``command_migration.json`` (loaded here), so the same recipes drive the Python,
TypeScript, and CLI SDKs and adding a command is a config entry, not code.

No CRIU: we sync the command's working dir + its on-disk resumable transcript
and re-invoke the command's own resume entrypoint in the VM. Works for any
command that (a) persists a transcript and (b) can resume from it.
"""
from __future__ import annotations

import glob as _glob
import json
import os
import re
import signal
import time
from pathlib import Path
from typing import Any, Optional

_CONFIG: Optional[dict] = None


def load_config() -> dict:
    """Load (and cache) the declarative command-migration recipe map."""
    global _CONFIG
    if _CONFIG is None:
        with open(Path(__file__).with_name("command_migration.json")) as fh:
            _CONFIG = json.load(fh)
    return _CONFIG


def _cwd_key(cwd: str) -> str:
    return re.sub(r"[^A-Za-z0-9]", "-", cwd)


def _subst(template: str, vars: dict[str, str]) -> str:
    out = template
    for k, v in vars.items():
        out = out.replace("${" + k + "}", v)
    return os.path.expanduser(out)


def _detect(argv: str, spec: dict) -> bool:
    if "argv_contains" in spec:
        return spec["argv_contains"] in argv
    if "argv_regex" in spec:
        return re.search(spec["argv_regex"], argv) is not None
    return False


def _find_session(spec: dict, vars: dict[str, str]) -> tuple[Optional[str], Optional[str]]:
    pattern = _subst(spec["glob"], vars)
    matches = sorted(_glob.glob(pattern, recursive=True), key=os.path.getmtime)
    if not matches:
        return None, None
    path = matches[-1] if spec.get("pick", "newest_mtime") == "newest_mtime" else matches[-1]
    idspec = spec.get("id", "stem")
    if idspec == "stem":
        sid = Path(path).stem
    elif idspec == "session_path":
        sid = path
    elif isinstance(idspec, dict) and "regex" in idspec:
        m = re.search(idspec["regex"], os.path.basename(path))
        sid = m.group(1) if m else Path(path).stem
    else:
        sid = Path(path).stem
    return path, sid


def discover(pid: int) -> dict:
    """Inspect /proc/<pid> and match it to a recipe. Returns the migration plan."""
    cfg = load_config()
    cwd = os.readlink(f"/proc/{pid}/cwd")
    argv = " ".join(
        c for c in open(f"/proc/{pid}/cmdline", "rb").read().decode("utf8", "replace").split("\0") if c
    ).lower()
    environ: dict[str, str] = {}
    for e in open(f"/proc/{pid}/environ", "rb").read().decode("utf8", "replace").split("\0"):
        if "=" in e:
            k, v = e.split("=", 1)
            environ[k] = v
    command = next((name for name, spec in cfg["commands"].items() if _detect(argv, spec["detect"])), None)
    session_path = session_id = None
    if command:
        vars = {"cwd": cwd, "cwd_key": _cwd_key(cwd)}
        session_path, session_id = _find_session(cfg["commands"][command]["session"], vars)
    return dict(cwd=cwd, command=command, environ=environ, session_path=session_path, session_id=session_id)


def quiesce(session_path: Optional[str], timeout: Optional[float] = None, stable_secs: Optional[float] = None) -> bool:
    """Wait until the transcript stops growing (in-flight turn flushed). True if it settled."""
    cfg = load_config().get("quiesce", {})
    timeout = timeout if timeout is not None else cfg.get("timeout_secs", 30)
    stable_secs = stable_secs if stable_secs is not None else cfg.get("stable_secs", 2.0)
    if not session_path or not os.path.exists(session_path):
        return True
    last, stable_at, t0 = -1, None, time.time()
    while time.time() - t0 < timeout:
        sz = os.path.getsize(session_path)
        if sz == last:
            if stable_at and time.time() - stable_at >= stable_secs:
                return True
        else:
            last, stable_at = sz, time.time()
        time.sleep(0.4)
    return False


def migrate(
    client,
    *,
    pid: int,
    source: str = "ubuntu-small",
    memory_mib: int = 2048,
    do_quiesce: bool = True,
    freeze_local: bool = False,
    kill_local: bool = False,
    probe: str = "In one short line: what were you last doing?",
    keys: Optional[dict[str, str]] = None,
):
    """Migrate the running command at ``pid`` into a fresh Arker VM and resume it.

    Returns (vm, resumed_output). Purely client-side: uses the SDK's fork + sync
    + run under the hood — no special server route.
    """
    cfg = load_config()
    info = discover(pid)
    command = info["command"]
    if command is None:
        raise ValueError(
            f"pid {pid} is not a recognized migratable command (no matching recipe in command_migration.json)"
        )
    spec = cfg["commands"][command]
    cwd, environ = info["cwd"], info["environ"]
    sid, spath = info["session_id"], info["session_path"]
    vars = {"cwd": cwd, "cwd_key": _cwd_key(cwd), "session_id": sid or "", "probe": json.dumps(probe)}

    if do_quiesce and not quiesce(spath):
        # completed turns are on disk; only the current unflushed turn is lost
        pass
    if freeze_local:
        os.kill(pid, signal.SIGSTOP)

    vm = client.fork(source, memory_mib=memory_mib)
    vm.run(f"mkdir -p {cwd}")
    vm.sync_dir(cwd, cwd.lstrip("/"))
    vm.run(spec["install"])

    # place the transcript
    if spath and os.path.exists(spath):
        place = spec["place"]
        if isinstance(place, dict) and "verbatim_under" in place:
            root = os.path.expanduser(place["from_host_root"])
            dest = place["verbatim_under"] + "/" + os.path.relpath(spath, root)
        else:
            dest = _subst(place, vars)
        vm.sync(dest, open(spath, "rb").read())

    # extra config files (resolve placeholders inside the JSON)
    for vm_path, body in spec.get("extra_files", {}).items():
        if "json" in body:
            raw = _subst(json.dumps(body["json"]), vars)
            vm.sync(_subst(vm_path, vars), raw.encode())

    # forward keys (explicit override wins over discovered environ) + recipe env
    key_env = keys or {k: environ[k] for k in spec.get("keys", []) if environ.get(k)}
    proc_env = dict(spec.get("env", {}))
    proc_env.update(key_env)

    sess = vm.create_session(env=proc_env, cwd=cwd)
    r = vm.run(_subst(spec["resume"], vars), session_id=sess.session_id, timeout=180)
    out = getattr(r, "stdout", b"") or b""
    out = out.decode("utf8", "replace") if isinstance(out, (bytes, bytearray)) else str(out)

    if kill_local:
        try:
            os.kill(pid, signal.SIGTERM)
        except Exception:
            pass
    return vm, out

File: typescript/src/migrate.ts
/**
 * Command migration: move a running command (by PID) into an Arker VM.
 *
 * This is the config-driven engine — every command-specific detail lives in
 * `command_migration.json` (bundled alongside this module and loaded here),
 * so the same recipes drive the Python, TypeScript, and CLI SDKs and adding a
 * command is a config entry, not code in three languages. Keep
 * `typescript/src/command_migration.json` byte-identical to the canonical
 * copy at `python/src/arker/command_migration.json` — `tests/migrate.ts`
 * diffs the two against each other when both are present on disk (i.e. in a
 * monorepo checkout) and fails the build if they drift.
 *
 * No CRIU: we sync the command's working dir + its on-disk resumable
 * transcript and re-invoke the command's own resume entrypoint in the VM.
 * Works for any command that (a) persists a transcript and (b) can resume
 * from it.
 *
 * Node-only: it reads `/proc`, the local filesystem, and sends POSIX signals
 * — the same constraint {@link VM.syncDir} already carries.
 */
import { existsSync, readFileSync, readdirSync, readlinkSync, statSync } from "node:fs";
import { homedir } from "node:os";
import { basename, join, relative } from "node:path";

import migrationConfigJson from "./command_migration.json";
import type { Arker, VM } from "./index.js";

// ── Recipe shape (mirrors command_migration.json's own `recipe_schema`) ────

export interface DetectSpec {
  argv_contains?: string;
  argv_regex?: string;
}

export type SessionIdSpec = "stem" | "session_path" | { regex: string };

export interface SessionSpec {
  glob: string;
  pick?: "newest_mtime" | "oldest_mtime";
  id?: SessionIdSpec;
}

export type PlaceSpec = string | { verbatim_under: string; from_host_root: string };

export interface CommandSpec {
  detect: DetectSpec;
  session: SessionSpec;
  requires?: Record<string, unknown>;
  install: string;
  place: PlaceSpec;
  extra_files?: Record<string, { json: unknown }>;
  env?: Record<string, string>;
  resume: string;
  keys?: string[];
}

export interface MigrationConfig {
  version: number;
  kind: string;
  description: string;
  quiesce?: { stable_secs?: number; timeout_secs?: number };
  commands: Record<string, CommandSpec>;
}

/** The parsed recipe map. esbuild/tsup inline the JSON at bundle time (and
 * Bun/tsc read it straight off disk under `bun test`/`tsc`), so there is no
 * runtime file lookup to get wrong across the ESM/CJS dual build. */
export function loadConfig(): MigrationConfig {
  return migrationConfigJson as unknown as MigrationConfig;
}

// ── Small pure helpers — exported for direct unit testing, mirroring the
// underscore-prefixed-but-still-imported helpers in python/src/arker/migrate.py ──

export function cwdKey(cwd: string): string {
  return cwd.replace(/[^A-Za-z0-9]/g, "-");
}

// Prefer $HOME over os.homedir(): Python's os.path.expanduser (what this
// mirrors) checks $HOME first and only falls back to the pwd database if it's
// unset, and some JS runtimes cache os.homedir() at process start rather than
// re-reading $HOME — which would make it deaf to a test (or a caller) that
// points HOME elsewhere for the lifetime of one call.
function expandHome(p: string): string {
  const home = process.env.HOME || homedir();
  if (p === "~") return home;
  if (p.startsWith("~/")) return join(home, p.slice(2));
  return p;
}

export function subst(template: string, vars: Record<string, string>): string {
  let out = template;
  for (const [k, v] of Object.entries(vars)) {
    out = out.split(`\${${k}}`).join(v);
  }
  return expandHome(out);
}

export function detectCommand(argv: string, spec: DetectSpec): boolean {
  if (spec.argv_contains !== undefined) return argv.includes(spec.argv_contains);
  if (spec.argv_regex !== undefined) return new RegExp(spec.argv_regex).test(argv);
  return false;
}

function stem(path: string): string {
  const base = basename(path);
  const dot = base.lastIndexOf(".");
  return dot > 0 ? base.slice(0, dot) : base;
}

/**
 * Minimal glob supporting exactly what the recipes need: `*` (any chars
 * within one path segment) and `**` (zero or more directories), matching
 * Python's `glob.glob(pattern, recursive=True)` semantics for these two
 * wildcards. No brace expansion, no `?`, no character classes — recipes
 * don't use them, and by the time a pattern reaches here every `${...}`
 * placeholder has already been substituted with a literal (non-wildcard)
 * value, so there's nothing else to support.
 */
function globSync(pattern: string): string[] {
  const expanded = expandHome(pattern);
  const isAbsolute = expanded.startsWith("/");
  const segments = expanded.split("/").filter((s) => s.length > 0);
  const results: string[] = [];

  function segmentRegex(seg: string): RegExp {
    const escaped = seg.replace(/[.+^${}()|[\]\\]/g, "\\$&").replace(/\*/g, ".*");
    return new RegExp(`^${escaped}$`);
  }

  function walk(base: string, segIdx: number): void {
    if (segIdx === segments.length) {
      results.push(base);
      return;
    }
    const seg = segments[segIdx]!;
    let entries: string[];
    try {
      entries = readdirSync(base);
    } catch {
      return;
    }
    if (seg === "**") {
      // `**` matches zero directories too: try the rest of the pattern here...
      walk(base, segIdx + 1);
      // ...and one-or-more, by recursing into every subdirectory while still
      // consuming `**`.
      for (const entry of entries) {
        const full = join(base, entry);
        let isDir: boolean;
        try {
          isDir = statSync(full).isDirectory();
        } catch {
          continue;
        }
        if (isDir) walk(full, segIdx);
      }
      return;
    }
    const re = segmentRegex(seg);
    const isLast = segIdx === segments.length - 1;
    for (const entry of entries) {
      if (!re.test(entry)) continue;
      const full = join(base, entry);
      if (isLast) {
        results.push(full);
        continue;
      }
      let isDir: boolean;
      try {
        isDir = statSync(full).isDirectory();
      } catch {
        continue;
      }
      if (isDir) walk(full, segIdx + 1);
    }
  }

  walk(isAbsolute ? "/" : ".", 0);
  return results;
}

export function findSession(
  spec: SessionSpec,
  vars: Record<string, string>,
): [path: string | null, sessionId: string | null] {
  const pattern = subst(spec.glob, vars);
  const matches = globSync(pattern)
    .map((p) => {
      try {
        return { p, mtime: statSync(p).mtimeMs };
      } catch {
        return null;
      }
    })
    .filter((m): m is { p: string; mtime: number } => m !== null)
    .sort((a, b) => a.mtime - b.mtime);
  if (matches.length === 0) return [null, null];
  // `matches` is sorted by mtime ascending; newest_mtime (default) is the
  // last entry, oldest_mtime the first.
  const chosen = spec.pick === "oldest_mtime" ? matches[0]! : matches[matches.length - 1]!;
  const idSpec = spec.id ?? "stem";
  let sessionId: string;
  if (idSpec === "stem") {
    sessionId = stem(chosen.p);
  } else if (idSpec === "session_path") {
    sessionId = chosen.p;
  } else if (typeof idSpec === "object" && "regex" in idSpec) {
    const m = new RegExp(idSpec.regex).exec(basename(chosen.p));
    sessionId = m ? m[1]! : stem(chosen.p);
  } else {
    sessionId = stem(chosen.p);
  }
  return [chosen.p, sessionId];
}

// ── /proc introspection ─────────────────────────────────────────────

export interface DiscoverResult {
  cwd: string;
  command: string | null;
  environ: Record<string, string>;
  sessionPath: string | null;
  sessionId: string | null;
}

/** Inspect `/proc/<pid>` and match it to a recipe. Returns the migration plan. */
export function discover(pid: number): DiscoverResult {
  const cfg = loadConfig();
  const cwd = readlinkSync(`/proc/${pid}/cwd`);
  const argv = readFileSync(`/proc/${pid}/cmdline`, "utf8")
    .split("\0")
    .filter((s) => s.length > 0)
    .join(" ")
    .toLowerCase();
  const environ: Record<string, string> = {};
  for (const entry of readFileSync(`/proc/${pid}/environ`, "utf8").split("\0")) {
    const eq = entry.indexOf("=");
    if (eq > -1) environ[entry.slice(0, eq)] = entry.slice(eq + 1);
  }
  const command =
    Object.entries(cfg.commands).find(([, spec]) => detectCommand(argv, spec.detect))?.[0] ?? null;
  let sessionPath: string | null = null;
  let sessionId: string | null = null;
  if (command !== null) {
    const vars = { cwd, cwd_key: cwdKey(cwd) };
    [sessionPath, sessionId] = findSession(cfg.commands[command]!.session, vars);
  }
  return { cwd, command, environ, sessionPath, sessionId };
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/** Wait until the transcript stops growing (in-flight turn flushed). True if
 * it settled, false if `timeoutSecs` elapsed first. */
export async function quiesce(
  sessionPath: string | null,
  timeoutSecs?: number,
  stableSecs?: number,
): Promise<boolean> {
  const cfg = loadConfig().quiesce ?? {};
  const timeout = (timeoutSecs ?? cfg.timeout_secs ?? 30) * 1000;
  const stable = (stableSecs ?? cfg.stable_secs ?? 2.0) * 1000;
  if (!sessionPath || !existsSync(sessionPath)) return true;
  let last = -1;
  let stableAt: number | null = null;
  const t0 = Date.now();
  while (Date.now() - t0 < timeout) {
    const size = statSync(sessionPath).size;
    if (size === last) {
      if (stableAt !== null && Date.now() - stableAt >= stable) return true;
    } else {
      last = size;
      stableAt = Date.now();
    }
    await sleep(400);
  }
  return false;
}

// ── migrate() ──────────────────────────────────────────────────────

export interface MigrateOptions {
  pid: number;
  source?: string;
  memoryMib?: number;
  doQuiesce?: boolean;
  freezeLocal?: boolean;
  killLocal?: boolean;
  probe?: string;
  keys?: Record<string, string>;
}

export interface MigrateResult {
  vm: VM;
  command: string;
  output: string;
}

/**
 * Migrate the running command at `pid` into a fresh Arker VM and resume it.
 *
 * Purely client-side: uses the SDK's fork + sync + syncDir + run under the
 * hood — no special server route. Mirrors `python/src/arker/migrate.py`'s
 * `migrate()` step for step.
 */
export async function migrate(client: Arker, options: MigrateOptions): Promise<MigrateResult> {
  const {
    pid,
    source = "ubuntu-small",
    memoryMib = 2048,
    doQuiesce = true,
    freezeLocal = false,
    killLocal = false,
    probe = "In one short line: what were you last doing?",
    keys,
  } = options;

  const cfg = loadConfig();
  const info = discover(pid);
  const { command, cwd, environ } = info;
  if (command === null) {
    throw new Error(
      `pid ${pid} is not a recognized migratable command (no matching recipe in command_migration.json)`,
    );
  }
  const spec = cfg.commands[command]!;
  const sessionId = info.sessionId;
  const sessionPath = info.sessionPath;
  const vars: Record<string, string> = {
    cwd,
    cwd_key: cwdKey(cwd),
    session_id: sessionId ?? "",
    probe: JSON.stringify(probe),
  };

  if (doQuiesce) {
    // Completed turns are on disk regardless of whether this settles before
    // its timeout; only the current unflushed turn is at risk either way, so
    // the result is intentionally not checked — same as the Python engine.
    await quiesce(sessionPath);
  }
  if (freezeLocal) {
    process.kill(pid, "SIGSTOP");
  }

  const vm = await client.fork(source, { resources: { memory_mib: memoryMib } });
  await vm.run(`mkdir -p ${cwd}`);
  await vm.syncDir(cwd, cwd.replace(/^\/+/, ""));
  await vm.run(spec.install);

  // Place the transcript.
  if (sessionPath && existsSync(sessionPath)) {
    const place = spec.place;
    let dest: string;
    if (typeof place === "object" && "verbatim_under" in place) {
      const root = expandHome(place.from_host_root);
      dest = `${place.verbatim_under}/${relative(root, sessionPath)}`;
    } else {
      dest = subst(place, vars);
    }
    await vm.sync(dest, readFileSync(sessionPath));
  }

  // Extra config files (resolve placeholders inside the JSON).
  for (const [vmPathTemplate, body] of Object.entries(spec.extra_files ?? {})) {
    if ("json" in body) {
      const raw = subst(JSON.stringify(body.json), vars);
      await vm.sync(subst(vmPathTemplate, vars), raw);
    }
  }

  // Forward keys (explicit override wins over discovered environ) + recipe env.
  //
  // NOTE: auto-discovery only ever forwards environment VARIABLES present in
  // the source process's /proc/<pid>/environ. Tools whose default login flow
  // stores credentials in a FILE instead (e.g. Claude Code's interactive
  // `claude login`, which writes ~/.claude/.credentials.json rather than
  // setting ANTHROPIC_API_KEY) are invisible to this mechanism no matter what
  // is added to a recipe's `keys` list — there is nothing in environ to find.
  // That case has no automatic path today; the caller must pass an explicit
  // `keys` override (e.g. a `claude setup-token`-minted CLAUDE_CODE_OAUTH_TOKEN)
  // to authenticate the resumed process in the VM.
  const keyEnv =
    keys ??
    Object.fromEntries((spec.keys ?? []).filter((k) => environ[k]).map((k) => [k, environ[k]!]));
  const procEnv = { ...(spec.env ?? {}), ...keyEnv };

  const session = await vm.createSession({ env: procEnv, cwd });
  const result = await vm.run(subst(spec.resume, vars), {
    session_id: session.session_id,
    timeout: 180,
  });
  const output = result.type === "completed" ? result.stdout : "";

  if (killLocal) {
    try {
      process.kill(pid, "SIGTERM");
    } catch {
      // Already gone — fine, that was the goal.
    }
  }

  return { vm, command, output };
}
